phip/clipped_factorization_analysis.py

```diff
@@ -0,0 +1,144 @@
+import time
+
+import pandas
+import numpy
+
+import tensorflow as tf
+from tensorflow.contrib.distributions import percentile
+
+
+def do_clipped_factorization_analysis(
+        counts_df,
+        rank=3,
+        truncate_percentile=99.9,
+        learning_rate=1.0,
+        minibatch_size=1024 * 32,
+        patience=5,
+        max_epochs=1000,
+        log_every_seconds=10):
+    """
+    Implements the factorization:
+
+    X = AB
+        where X is (n x s), A is (n x r), and B is (r x s)
+
+    by minimizing the clipped loss:
+
+        ||minimum(X - AB, percentile(X - AB, q)||_2 + unclipped_term
+
+    The minimum is taken elementwise, and ||M||_2 is the Frobenius norm.
+    q is a parameter giving the percentile to clip at, default value 99.9. This
+    truncation makes the factorization robust to outliers, which are our hits.
+
+    If the above is optimized without an `unclipped_term`, a few phage clones
+    may have all of their residuals above the truncation threshold. Once this
+    happens they will likely stay stuck there since they do not contribute to
+    the gradient. The `unclipped_term` fixes this by providing a small nudge
+    toward smaller errors without truncation. See implementation for details.
+    """
+
+    # Non-tf setup
+    reads_per_million = (counts_df * 1e6 / counts_df.sum(0)).astype("float32")
+    (n, s) = reads_per_million.shape
+
+    # Placeholders
+    target = tf.placeholder(
+        name="target", dtype="float32", shape=[None, s])
+    minibatch_indices = tf.placeholder(
+        name="minibatch_indices", dtype="int32")
+
+    # Variables
+    a = tf.Variable(
+        numpy.random.rand(n, rank), name="A", dtype="float32")
+    b = tf.Variable(
+        numpy.random.rand(rank, s), name="B", dtype="float32")
+    truncate_threshold = tf.Variable(reads_per_million.max().max())
+
+    # Derived quantities
+    reconstruction = tf.matmul(
+        tf.gather(a, minibatch_indices), b)
```

Comment:
Is there an implication for sampling rows but not cols? Or is this basically just a pragmatic issue?

Reply:
Yeah taking the minibatches row-wise is really just a pragmatic thing, since that's the long dimension in our data